src/pow.rs

```diff
@@ -0,0 +1,1314 @@
+        // with it. The exponent needs to have 3 subtracted from it, hence
+        // this goofy decoding code. 3 is due to 3 bytes in the mantissa.
+        let (mant, expt) = {
+            let unshifted_expt = bits >> 24;
+            if unshifted_expt <= 3 {
+                ((bits & 0xFFFFFF) >> (8 * (3 - unshifted_expt as usize)), 0)
+            } else {
+                (bits & 0xFFFFFF, 8 * ((bits >> 24) - 3))
+            }
+        };
+
+        // The mantissa is signed but may not be negative
+        if mant > 0x7FFFFF {
+            Target::default()
+        } else {
+            Target::from(mant as u64) << (expt as usize)
+        }
+    }
+
+    /// Computes the compact value from a [`Target`] representation.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use bitcoin::Target;
+    /// assert_eq!(0x1d00ffff, Target::to_compact(&Target::from_compact(0x1d00ffff)));
+    /// ```
+    pub fn to_compact(&self) -> u32 {
+        let mut size = (self.0.bits() + 7) / 8;
+        let mut compact = if size <= 3 {
+            (self.0.low_u64() << (8 * (3 - size))) as u32
+        } else {
+            let bn = *self >> (8 * (size - 3));
+            bn.0.low_u32()
+        };
+
+        if (compact & 0x00800000) != 0 {
+            compact >>= 8;
+            size += 1;
+        }
+
+        compact | (size << 24) as u32
+    }
+}
+
+do_impl!(Target);
+
+impl_mul!(Target, u8);
+impl_mul!(Target, u16);
+impl_mul!(Target, u32);
+
+impl_div!(Target, u8);
+impl_div!(Target, u16);
+impl_div!(Target, u32);
+
+impl_shift!(Target, u8);
+impl_shift!(Target, u16);
+impl_shift!(Target, u32);
+impl_shift!(Target, usize);
+
+impl Div for Target {
+    type Output = Self;
+    fn div(self, rhs: Self) -> Self { Self(self.0 / rhs.0) }
+}
+
+/// Little-endian 256 bit integer type.
+#[derive(Copy, Clone, PartialEq, Eq, Hash, Default)]
+struct Uint256(pub [u64; 4]);
+
+impl Uint256 {
+    const MAX: Uint256 = Uint256([
+        0xffff_ffff_ffff_ffff_u64,
+        0xffff_ffff_ffff_ffff_u64,
+        0xffff_ffff_ffff_ffff_u64,
+        0xffff_ffff_ffff_ffff_u64,
+    ]);
+
+    const ZERO: Uint256 = Uint256([0, 0, 0, 0]);
+
+    const ONE: Uint256 = Uint256([1, 0, 0, 0]);
+
+    /// Calculates 2^256 / (x + 1) where x is a 256 bit unsigned integer.
+    ///
+    /// 2**256 / (x + 1) == ~x / (x + 1) + 1
+    ///
+    /// (Equation shamelessly stolen from bitcoind)
+    fn inverse(&self) -> Uint256 {
+        // We should never have a target/work of zero so this doesn't matter
+        // that much but we define the inverse of 0 as MAX.
+        if *self == Uint256::ZERO {
+            return Uint256::MAX;
+        }
+
+        // We define the inverse of max as 1.
+        if *self == Uint256::MAX {
+            return Uint256::ONE;
+        }
+
+        // And the inverse of 1 as max.
+        if *self == Uint256::ONE {
+            return Uint256::MAX;
+        }
+
+        let mut denom = *self;
+        denom.increment();
```

Comment:
Perhaps rename it to `wrapping_inc_assign()`?

Reply:
Agree, done.